packages/@aws-cdk/aws-kinesisanalyticsv2/lib/flink-application.ts

```diff
@@ -0,0 +1,328 @@
+      public readonly role = undefined;
+      public readonly grantPrincipal = new iam.UnknownPrincipal({ resource: this });
+
+      public readonly applicationName = attributes.applicationName;
+      public readonly applicationArn = attributes.applicationArn;
+    }
+
+    return new Import(scope, id);
+  }
+
+  public readonly applicationArn: string;
+  public readonly applicationName: string;
+
+  // Role must be optional for JSII compatibility
+  public readonly role?: iam.IRole;
+
+  public readonly grantPrincipal: iam.IPrincipal;
+
+  constructor(scope: Construct, id: string, props: FlinkApplicationProps) {
+    super(scope, id, { physicalName: props.applicationName });
+    validateFlinkApplicationProps(props);
+
+    this.role = props.role ?? new iam.Role(this, 'Role', {
+      assumedBy: new iam.ServicePrincipal('kinesisanalytics.amazonaws.com'),
+    });
+    this.grantPrincipal = this.role;
+
+    // Permit metric publishing to CloudWatch
+    this.role.addToPrincipalPolicy(new iam.PolicyStatement({
+      actions: ['cloudwatch:PutMetricData'],
+      resources: ['*'],
+    }));
+
+    // TODO: Refactor to make this bug impossible
+    const code = props.code.bind(this);
+    if (!props.code.bucket) {
+      throw new Error('BUG: Bound code is missing its bucket.');
+    }
+    props.code.bucket.grantRead(this);
+
+    const resource = new CfnApplication(this, 'Resource', {
+      runtimeEnvironment: props.runtime.value,
+      serviceExecutionRole: this.role.roleArn,
+      applicationConfiguration: {
+        applicationCodeConfiguration: code,
+        environmentProperties: props.propertyGroups?.length
+          ? { propertyGroups: props.propertyGroups.map(pg => pg.toCfn()) }
+          : undefined,
+        flinkApplicationConfiguration: flinkApplicationConfiguration({
+          checkpointingEnabled: props.checkpointingEnabled,
+          checkpointInterval: props.checkpointInterval,
+          minPauseBetweenCheckpoints: props.minPauseBetweenCheckpoints,
+          logLevel: props.logLevel,
+          metricsLevel: props.metricsLevel,
+          autoScalingEnabled: props.autoScalingEnabled,
+          parallelism: props.parallelism,
+          parallelismPerKpu: props.parallelismPerKpu,
+        }),
+        applicationSnapshotConfiguration: {
+          snapshotsEnabled: props.snapshotsEnabled ?? true,
+        },
+      },
+    });
+    resource.node.addDependency(this.role);
+
+    const logGroup = new logs.LogGroup(this, 'LogGroup', {
+      retention: logs.RetentionDays.ONE_WEEK,
+      removalPolicy: core.RemovalPolicy.DESTROY,
+    });
+
+    const logStream = new logs.LogStream(this, 'LogStream', {
+      logGroup,
+      removalPolicy: core.RemovalPolicy.DESTROY,
+    });
+
+    const logStreamArn = `arn:${core.Aws.PARTITION}:logs:${core.Aws.REGION}:${core.Aws.ACCOUNT_ID}:log-group:${logGroup.logGroupName}:log-stream:${logStream.logStreamName}`;
```

Comment:
Added config for this resolving open question 3. I ended up just flattening the config and appending it to the `FlinkApplicationProps` type. When I introduced a new type `ka.Logging` I realized that it seems to add little value and it made config discoverability worse. It also make `logLevel` confusing because it would not have been part of the `ka.Logging` type. Now every prop relevant to logging begins with `log`.